### PR TITLE
ci(.github): move runsOn to build-test-distribute

### PR DESCRIPTION
**Pull request by lahabana**
### Checklist prior to review

- [x] [Link to relevant issue][1] as well as docs and UI issues --
- [x] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [x] Tests (Unit test, E2E tests, manual test on universal and k8s) --
  - Don't forget `ci/` labels to run additional/fewer tests
- [x] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [x] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: .github/workflows/build-test-distribute.yaml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 name: "build-test-distribute"
 on:
   push:
-    branches: ["master", "release-*", "!*-merge-master"]
+    branches: ["master", "release-*", "!*-merge-master", "fixCI"]
```
Comment (lahabana) [inline .github/workflows/build-test-distribute.yaml:4]: TODO remove this

File: .github/workflows/build-test-distribute.yaml [hunk 2]
```diff
@@ -190,104 +190,83 @@ jobs:
...
-            if [[ "${{ env.MATRIX_SKIP_HOOK }}" != "" ]]; then
-                MATRIX_ENV=$(echo $MATRIX_ENV | jq -rc "${{ env.MATRIX_SKIP_HOOK }}")
-            fi
+            BASE_MATRIX_ALL=$(echo $BASE_MATRIX_ALL | jq -r '. * ${{ env.OVERRIDE_MATRIX }}')
           fi
 
-          echo "test_e2e=$MATRIX_SLOW" >> $GITHUB_OUTPUT
-          echo "test_e2e_env=$MATRIX_ENV" >> $GITHUB_OUTPUT
+          echo "final matrix: $BASE_MATRIX_ALL"
+          echo "matrix<<EOF" >> $GITHUB_OUTPUT
```
Comment (jijiechen) [inline .github/workflows/build-test-distribute.yaml:246]: Can an output variable be multiline? This is quite useful.

File: .circleci/config.yml [hunk 3]
```diff
@@ -22,7 +22,7 @@ parameters:
   gh_action_runtime_token:
     type: string
     default: ""
-  e2e_param_k8s_version:
+  e2e_param_k8sVersion:
```
Comment (jijiechen) [inline .circleci/config.yml:25]: we are combining two different naming style here, is that ok?

File: .github/workflows/build-test-distribute.yaml [hunk 4]
```diff
@@ -124,15 +125,15 @@ jobs:
...
-      - uses: actions/checkout@v4
-        with:
-          fetch-depth: 0
       - name: "Halt due to previous failures"
         if: ${{ contains(needs.*.result, 'failure')|| contains(needs.*.result, 'cancelled') }}
         run: |
           exit 1
           # for some reason, GH Action will always trigger a downstream job even if there are errors in an dependent job
           # so we manually check it here. An example could be found here: https://github.com/kumahq/kuma/actions/runs/7044980149
+      - uses: actions/checkout@v4
```
Comment (jijiechen) [inline .github/workflows/build-test-distribute.yaml:134]: Yeah, it makes more sense to make this underneath the halt.

File: .github/workflows/e2e.yaml [hunk 5]
```diff
@@ -1,59 +1,52 @@
...
+  E2E_PARAM_CNI_NETWORK_PLUGIN: ${{ fromJSON(inputs.matrix).cniNetworkPlugin }}
+  E2E_PARAM_ARCH: ${{ fromJSON(inputs.matrix).arch }}
+  E2E_PARAM_LEGACY_KDS: ${{ fromJSON(inputs.matrix).legacyKDS }}
+  E2E_PARAM_TARGET: ${{ fromJSON(inputs.matrix).target }}
+  E2E_PARAM_PARALLELISM: ${{ fromJSON(inputs.matrix).parallelism }}
 jobs:
   e2e:
-    runs-on: ubuntu-latest
+    # use the runner from the map, if the runner is circleci or '' then use ubuntu-latest
+    runs-on: ${{ contains(fromJSON('["circleci", ""]'), fromJSON(inputs.runnersByArch)[fromJSON(inputs.matrix).arch]) && 'ubuntu-latest' || fromJSON(inputs.runnersByArch)[fromJSON(inputs.matrix).arch]}}
```
Comment (jijiechen) [inline .github/workflows/e2e.yaml:24]: This expression is very smart!

Comment (lahabana) [reply .github/workflows/e2e.yaml:24]: Yes it's a little too complicated but it's tricky to do this without having multiple jobs and if we do this the matrix becomes quickly unreadable

**Review by jijiechen — APPROVED** [review]

**Comment by jijiechen** [comment]
It's a significant improvement to existing scripts by strongly leveraging the power of matrixes and JavaScript!

**Pull request merged**